src/server/ua_services_attribute.c

```diff
@@ -1402,6 +1415,51 @@ __UA_Server_write(UA_Server *server, const UA_NodeId *nodeId,
     return UA_Server_write(server, &wvalue);
 }
 
+#ifdef UA_ENABLE_HISTORIZING
+
+#include "ua_plugin_history_data_service.h"
+
+void
+Service_HistoryRead(UA_Server *server,
+                    UA_Session *session,
+                    const UA_HistoryReadRequest *request,
+                    UA_HistoryReadResponse *response) {
+    UA_HistoryReadResponse_init(response);
+    if (request->historyReadDetails.encoding != UA_EXTENSIONOBJECT_DECODED) {
+        response->responseHeader.serviceResult = UA_STATUSCODE_BADNOTSUPPORTED;
+        return;
+    }
+    if (request->historyReadDetails.content.decoded.type == &UA_TYPES[UA_TYPES_READRAWMODIFIEDDETAILS]) {
+        UA_ReadRawModifiedDetails * details = (UA_ReadRawModifiedDetails*)request->historyReadDetails.content.decoded.data;
+        if (details->isReadModified) {
+            // TODO add server->config.historyReadService.read_modified
+            response->responseHeader.serviceResult = UA_STATUSCODE_BADHISTORYOPERATIONUNSUPPORTED;
+            return;
+        } else {
+            if (server->config.historyDataService.readRaw) {
+                response->resultsSize = request->nodesToReadSize;
+
+                response->results = (UA_HistoryReadResult*)UA_Array_new(response->resultsSize, &UA_TYPES[UA_TYPES_HISTORYREADRESULT]);
+                server->config.historyDataService.readRaw(server->config.historyDataService.context,
+                                                           server,
+                                                           session ? &session->sessionId : NULL,
+                                                           session ? session->sessionHandle : NULL,
+                                                           request,
+                                                           details,
+                                                           response,
+                                                           response->results);
+                return;
```

Comment:
As the plugin functions kind of implement the service - if you want - the functions could return a operation level status code, which is used as service result. I'm not sure how important / useful this is.

Well, you already force it by accessing the `response` object, but it might be more convenient.

Reply:
The return statuscode is in the response. I do not see the requirement for an extra statuscode. What to do with that statuscode ? how should we react if this statuscode is bad and in the response is a bad statuscode and the other way around. If we have a bad statuscode we still need to answer the client. I like a singe source of true...

Reply:
The idea was to take the returned value as `serviceResult`. Sure you can set it in the response itself. Just wanted to discuss this, as it came into my mind on the first time digging through the PR. It's not exactly a requirement. :)